File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.cc
#include <bits/stdc++.h>
using namespace std;

vector<int> adj[1002];
bool visited[1002];

int main() {
	cin.tie(0) -> sync_with_stdio(false);

    int n, m; cin >> n >> m;
    while (m-- > 0) {
        int u, v; cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    int cnt = 0;
    for (int i = 1; i <= n; i++) {
        if (visited[i]) continue;
        queue<int> q;
        q.push(i);
        visited[i] = true;

        while (!q.empty()) {
            auto cur = q.front();
            q.pop();

            for (int nxt : adj[cur]) {
                if (visited[nxt]) continue;
                q.push(nxt);
                visited[nxt] = true;
            }
        }
        cnt++;
    }
    cout << cnt;
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.*;
import java.util.*;

class Main {

    static int n;
    static int m;
    static boolean[] visited = new boolean[1002];
    static List<List<Integer>> graph = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] parts = br.readLine().split(" ");
        n = Integer.parseInt(parts[0]);
        m = Integer.parseInt(parts[1]);
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        while (m-- > 0) {
            parts = br.readLine().split(" ");
            int u = Integer.parseInt(parts[0]);
            int v = Integer.parseInt(parts[1]);
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        int cnt = 0;
        for (int i = 1; i <= n; i++) {
            if (visited[i]) continue;
            bfs(i);
            cnt++;
        }

        System.out.print(cnt);
    }

    private static void bfs(int start) {
        Queue<Integer> queue = new ArrayDeque<>();
        queue.offer(start);
        visited[start] = true;

        while (!queue.isEmpty()) {
            var cur = queue.poll();

            for (int nxt : graph.get(cur)) {
                if (visited[nxt]) continue;
                queue.offer(nxt);
                visited[nxt] = true;
            }
        }
    }
}